### ISSUE TITLE
Source Routing

### ISSUE THREAD

**Issue by johandc** [libcsp/libcsp#28]
Before venturing into the great worlds of routing protocols, the simple tool of source routing should be added to CSP. This ensures that as long as nodes are informed of their neighboring nodes, a packet can always be successfully routed with manual source routing. This could be used as an emergency tool in order to modify routing tables that are invalid manually.

The implementation is quite simple, and not something that would interfere with normal CSP traffic. I suggest that we add a CSP_FSRCRTE flag to the header and add an optional routing table at the end of the data, just as we add other fields at the end. 

When transmitting: The routing table could consist of a 8-bit length field and a variable length array of 8-bit fields which specify the next hop address. Each intermediate node will find its own address in the table, and send the message to the next hop node. When the message arrives at the CSP end-node, there are a few options

When receiving: If a message arrived to the node by source routing, we can assume that the default route on the receiving node is no longer valid, or there may be some other reason not to use the default route. Therefore the receiving node should either remember the source route on a connection basis, and/or update it's routing table to the correct default route. Maybe there should be a "learn" flag on the source routed message, that would make all CSP-nodes update their routing table. In this case source routing could be used as a tool to update the routing tables in the network by simply sending a ping message to a specific node, using source routing and the "learn" flag.

For the first implementation i think that the connection handler should remember the incoming source route, and then reverse it for replies on the connection. Only one problem. Since all connection data is static, how do we get memory allocation for an arbitrary length source route?

**Comment by jledet** [comment]
I don't know about this one. Maybe we should just start by allowing multiple routes to destinations and use some kind of (manual) metric to decide between them.

**Comment by johandc** [comment]
The point is that we would like to be able to alter routing tables so that the network is somewhat reconfigurable (as long as there are no node address conflicts).

I recently added a CMP service to be able to update routing tables. Just like a remote csp_route_set() call. This could be used on a hop-by hop basis to configure the entire network from only one node. The source routing would not be needed then, since the routes could be updated by the operator manually.

Let's say we have a situation where all routing tables are completely empty, and we are sitting on the ground station and wants to contact the OBC on the satellite over the radio link. We would start by first defining a route locally for the TNC over KISS. Then we would send a route change request to the TNC in order to know the reverse route to the PC. Then we would send a ping message to the TNC to verify that the reverse route worked. The reason why we can do this is that no reverse route is needed in order to update the routing table. As long as messages can be transferred forwards we can always configure the backwards route. Okay so the next step would be the radio. So we will configure a forward route to the radio over the TNC's space link, and on the PC towards the TNC. Then we will send a route update to the COM system and configure the reverse route. And this then continues for all nodes in the network. Now we have successfully reconfigured the network, from scratch.

The original idea for this issue was to be able to specify the route directly in the message, and let it hop all the way to the end node, and letting the intermediate nodes automatically learn and update both the forwards and backwards route from the message. However this would only work if all the neighboring routes were configured correctly, and could never work from a completely corrupted or empty routing table. So the reconfiguration requirements was not fulfilled by this idea anyways.

Therefore i'm closing this now.
Fixed by #8133935
